### PR TITLE
Add support for classmethods and staticmethods in add_method

### LINKED ISSUE #7301: Plugin add_classmethod
**Feature request:**
A function like `mypy.plugins.common.add_method`, but with support for classmethods.

-----

More context:

I am working on a plugin, and would like to be able to detect and raise errors when an inherited (factory) classmethod is called, but *only* if certain conditions hold for the type (which I am computing in the `base_model_hook` and storing on `ctx.cls.info.metadata`).

The method is inherited from a parent class, and when the method hook is called it only contains references to the parent class, not the calling class (on which I've stored the metadata). (The parent does not have the appropriate type-specific metadata I want to access.)

In order to get around this, I had the idea to directly add a classmethod to the subclass so that it could trigger the method_hook itself. (Note: I'm not sure whether this would work since I'd be generating the method in the `base_class_hook`; I'm not sure whether it would get picked up by the `get_method_hook` or not.)

**(Any insight into how I can create a conditionally-erroring classmethod, where the error occurs based on data computed in the `base_model_hook`, would be appreciated.)**

Whether or not the above approach would work for this purpose, I didn't see an easy way to add a classmethod (similarly to how `add_method` works), and I think this might be a nice capability to suppport.

(This came up in a discussion with @ilevkivskyi on gitter, and he suggested I create an issue.)

### PR DESCRIPTION
**Pull request by svalentin**
### Description

Fixes #7301

Adds `is_classmethod` and `is_staticmethod` options to `add_method` so plugins can create class/static methods.
Includes tests.

The old PR #7796 didn't have tests and is stale for a long time.

## Test Plan
`pytest -k testAddClassMethodPlugin`
`pytest -k testIncrementalAddClassMethodPlugin`

### CODE REVIEW

File: mypy/plugins/common.py [hunk 1]
```diff
@@ -139,13 +150,21 @@ def add_method_to_class(
...
             cls.defs.body.remove(sym.node)
 
-    self_type = self_type or fill_typevars(info)
     if isinstance(api, SemanticAnalyzerPluginInterface):
         function_type = api.named_type("builtins.function")
     else:
         function_type = api.named_generic_type("builtins.function", [])
 
-    args = [Argument(Var("self"), self_type, None, ARG_POS)] + args
+    if is_classmethod:
```
Comment (sobolevn) [inline mypy/plugins/common.py:158]: I think that we need to `assert` that `is_classmethod` and `is_staticmethod` are not both `True` at the same time.

Comment (svalentin) [reply mypy/plugins/common.py:160]: Ah, yes. Added!

**Comment by svalentin** [comment]
Sorry, I had forgotten to sort imports. All tests should pass now!

File: test-data/unit/check-custom-plugin.test [hunk 2]
```diff
@@ -991,3 +991,20 @@ class Cls:
...
+
+[case testAddClassMethodPlugin]
+# flags: --config-file tmp/mypy.ini
+class BaseAddMethod: pass
+
+class MyClass(BaseAddMethod):
+    pass
+
+my_class = MyClass()
+reveal_type(MyClass.foo_classmethod())  # N: Revealed type is "None"
```
Comment (sobolevn) [inline test-data/unit/check-custom-plugin.test:1003]: Let's add `reveal_type(MyClass.foo_classmethod)` and `reveal_type(MyClass.foo_staticmethod)`.

This check would ensure that our signatures are correct.

File: test-data/unit/check-custom-plugin.test [hunk 3]
```diff
@@ -991,3 +991,20 @@ class Cls:
...
+reveal_type(my_class.foo_classmethod())  # N: Revealed type is "None"
+reveal_type(MyClass.foo_staticmethod(1))  # N: Revealed type is "builtins.str"
+reveal_type(my_class.foo_staticmethod(2))  # N: Revealed type is "builtins.str"
+my_class.foo_staticmethod("asdf")  # E: Argument 1 to "foo_staticmethod" of "MyClass" has incompatible type "str"; expected "int"
```
Comment (sobolevn) [inline test-data/unit/check-custom-plugin.test:1007]: Afterwards, we can remove this check.

File: test-data/unit/plugins/add_classmethod.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class ClassMethodPlugin(Plugin):
+    def get_base_class_hook(self, fullname: str):
+        if "BaseAddMethod" in fullname:
+            return add_extra_methods_hook
+        return None
+
+
+def add_extra_methods_hook(ctx):
```
Comment (sobolevn) [inline test-data/unit/plugins/add_classmethod.py:14]: Can you please type all arguments / returns? This is just a style we follow in other test plugins.

**Review by sobolevn — COMMENTED** [review]
Looking good! Thank you

File: test-data/unit/check-custom-plugin.test (hunk 2 above)
Comment (svalentin) [reply test-data/unit/check-custom-plugin.test:1003]: I can add the reveal_types, but I honestly would still leave the other reveal_types that check calling the function works.

**Comment by svalentin** [comment]
Please take another look!

File: test-data/unit/check-custom-plugin.test [hunk 5]
```diff
@@ -991,3 +991,21 @@ class Cls:
...
+class MyClass(BaseAddMethod):
+    pass
+
+my_class = MyClass()
+reveal_type(MyClass.foo_classmethod)  # N: Revealed type is "def ()"
+reveal_type(MyClass.foo_staticmethod)  # N: Revealed type is "def (builtins.int) -> builtins.str"
+reveal_type(MyClass.foo_classmethod())  # N: Revealed type is "None"
+reveal_type(my_class.foo_classmethod())  # N: Revealed type is "None"
+reveal_type(MyClass.foo_staticmethod(1))  # N: Revealed type is "builtins.str"
+reveal_type(my_class.foo_staticmethod(2))  # N: Revealed type is "builtins.str"
```
Comment (sobolevn) [inline test-data/unit/check-custom-plugin.test:1008]:
```suggestion
```

File: test-data/unit/check-incremental.test [hunk 6]
```diff
@@ -5911,6 +5911,39 @@ tmp/c.py:4: note: Revealed type is "TypedDict('a.N', {'r': Union[TypedDict('b.M'
...
+import a
+
+[file b.py.2]
+import a
+
+my_class = a.MyClass()
+reveal_type(a.MyClass.foo_classmethod())
+reveal_type(my_class.foo_classmethod())
+reveal_type(a.MyClass.foo_staticmethod(1))
+reveal_type(my_class.foo_staticmethod(2))
```
Comment (sobolevn) [inline test-data/unit/check-incremental.test:5938]:
```suggestion
reveal_type(MyClass.foo_classmethod)
reveal_type(MyClass.foo_staticmethod)
```

File: test-data/unit/check-incremental.test [hunk 7]
```diff
@@ -5911,6 +5911,39 @@ tmp/c.py:4: note: Revealed type is "TypedDict('a.N', {'r': Union[TypedDict('b.M'
...
+
+[rechecked b]
+[out2]
+tmp/b.py:4: note: Revealed type is "None"
+tmp/b.py:5: note: Revealed type is "None"
+tmp/b.py:6: note: Revealed type is "builtins.str"
+tmp/b.py:7: note: Revealed type is "builtins.str"
```
Comment (sobolevn) [inline test-data/unit/check-incremental.test:5945]:
```suggestion
tmp/b.py:4: note: Revealed type is "def ()"
tmp/b.py:5: note:  Revealed type is "def (builtins.int) -> builtins.str"
```

File: test-data/unit/check-incremental.test [hunk 8]
```diff
@@ -5911,6 +5911,35 @@ tmp/c.py:4: note: Revealed type is "TypedDict('a.N', {'r': Union[TypedDict('b.M'
...
+
+my_class = a.MyClass()
+reveal_type(MyClass.foo_classmethod)
+reveal_type(MyClass.foo_staticmethod)
+
+[rechecked b]
+[out2]
+tmp/b.py:4: note: Revealed type is "def ()"
+tmp/b.py:5: note:  Revealed type is "def (builtins.int) -> builtins.str"
+
```
Comment (sobolevn) [inline test-data/unit/check-incremental.test:5942]:
```suggestion
my_class = a.MyClass()
reveal_type(a.MyClass.foo_classmethod)
reveal_type(a.MyClass.foo_staticmethod)
reveal_type(my_class.foo_classmethod)
reveal_type(my_class.foo_staticmethod)

[rechecked b]
[out2]
tmp/b.py:4: note: Revealed type is "def ()"
tmp/b.py:5: note:  Revealed type is "def (builtins.int) -> builtins.str"
tmp/b.py:6: note: Revealed type is "def ()"
tmp/b.py:7: note:  Revealed type is "def (builtins.int) -> builtins.str"
```

Comment (sobolevn) [reply test-data/unit/check-incremental.test:5942]: Sorry, I didn't get it right the first time :)

File: test-data/unit/check-incremental.test [hunk 9]
```diff
@@ -5911,6 +5911,38 @@ tmp/c.py:4: note: Revealed type is "TypedDict('a.N', {'r': Union[TypedDict('b.M'
...
+reveal_type(a.MyClass.foo_staticmethod)
+reveal_type(my_class.foo_classmethod)
+reveal_type(my_class.foo_staticmethod)
+
+[rechecked b]
+[out2]
+tmp/b.py:4: note: Revealed type is "def ()"
+tmp/b.py:5: note:  Revealed type is "def (builtins.int) -> builtins.str"
+tmp/b.py:6: note: Revealed type is "def ()"
+tmp/b.py:7: note:  Revealed type is "def (builtins.int) -> builtins.str"
```
Comment (sobolevn) [inline test-data/unit/check-incremental.test:5945]:
```suggestion
tmp/b.py:4: note: Revealed type is "def ()"
tmp/b.py:5: note: Revealed type is "def (builtins.int) -> builtins.str"
tmp/b.py:6: note: Revealed type is "def ()"
tmp/b.py:7: note: Revealed type is "def (builtins.int) -> builtins.str"
```

🤦 🤦 😄

**Review by sobolevn — APPROVED** [review]

**Comment by svalentin** [comment]
Btw, please remember to merge this!. "Only those with write access to this repository can merge pull requests".

**Comment by sobolevn** [comment]
Can you please resolve merge conflicts?

**Comment by svalentin** [comment]
merge conflicts resolved

**Review by sobolevn — APPROVED** [review]

**Pull request merged**